core/src/commonMain/kotlin/com/ustadmobile/core/contentjob/ContentJobRunner.kt

```diff
@@ -0,0 +1,274 @@
+            close()
+        }
+    }
+
+    private fun CoroutineScope.launchProcessor(id: Int, channel: ReceiveChannel<ContentJobItemAndContentJob>) = launch {
+        val tmpDir = createTemporaryDir("job-$id")
+
+        for(item in channel) {
+            val processContext = ProcessContext(tmpDir, mutableMapOf())
+            println("Proessor #$id processing job #${item.contentJobItem?.cjiUid} attempt #${item.contentJobItem?.cjiAttemptCount}")
+
+            var processResult: ProcessResult? = null
+            var processException: Throwable? = null
+            var mediatorObserver: DoorObserver<Pair<Int, Boolean>>?= null
+            val mediatorLiveData = JobConnectivityLiveData(connectivityLiveData,
+                    db.contentJobDao.findMeteredAllowedLiveData(
+                            item.contentJob?.cjUid ?: 0))
+
+            try {
+
+                item.contentJobItem?.cjiUid?.let { db.contentJobItemDao.updateStartTimeForJob(it, getSystemTimeInMillis()) }
+
+                val sourceUri = item.contentJobItem?.sourceUri?.let { DoorUri.parse(it) }
+                    ?: throw IllegalArgumentException("ContentJobItem #${item.contentJobItem?.cjiUid} has no source uri!")
+
+                var metadataResult: MetadataResult? = null
+                if(item.contentJobItem?.cjiContentEntryUid == 0L) {
+                    metadataResult = contentPluginManager.extractMetadata(sourceUri, processContext)
+                        ?: throw FatalContentJobException("ContentJobItem #${item.contentJobItem?.cjiUid}: cannot extract metadata")
+                    val contentEntryUid = repo.contentEntryDao.insertAsync(metadataResult.entry)
+                    item.contentJobItem?.cjiContentEntryUid = contentEntryUid
+                    db.contentJobItemDao.updateContentEntryUid(item.contentJobItem?.cjiUid ?: 0,
+                        contentEntryUid)
+
+                    if(item.contentJobItem?.cjiParentContentEntryUid != 0L){
+                        ContentEntryParentChildJoin().apply {
+                            cepcjParentContentEntryUid = item.contentJobItem?.cjiParentContentEntryUid ?: 0L
+                            cepcjChildContentEntryUid = item.contentJobItem?.cjiContentEntryUid ?: 0L
+                            cepcjUid = repo.contentEntryParentChildJoinDao.insert(this)
+                        }
+                    }
+                }
+
+                val pluginId = if(item.contentJobItem?.cjiPluginId == 0) {
+                    metadataResult?.pluginId ?: contentPluginManager.extractMetadata(sourceUri,
+                        processContext)?.pluginId ?:
+                        throw FatalContentJobException("ContentJobItem #${item.contentJobItem?.cjiUid}: cannot determine pluginId")
+                }else {
+                    item.contentJobItem?.cjiPluginId ?: 0
+                }
+
+                val plugin = contentPluginManager.getPluginById(pluginId)
+
+                val job = async {
```

Comment:
I think this should use the deferred result e.g.

```
val jobResult = async {
   plugin.processJob(item, processContext, this@ContentJobRunner)
}
```

then val processResult = jobResult.await()

That's cleaner than doing the assignment inside the async.